Fix compatibility with maven artifacts

garc33 opened the pull request:
- make generateUndoundJar task accessible from project config
  example : 
  artifacts{
  archives generateUnboundTask
  }
- add classifier to be able to upload generated jar into a repo

garc33 committed:
Fix compatibility with maven artifacts

* make generateUndoundJar task accessible from project config
example : 
artifacts{
	archives generateUnboundTask
}

* add classifier to be able to upload generated jar into a repo

bcorne commented on src/main/groovy/com/ullink/gradle/plugins/jibx/JIBXPlugin.groovy:
```diff
@@ -17,6 +17,14 @@ class JIBXPlugin implements Plugin<Project> {
+            }
+            from project.sourceSets.main.output.files
+        }
+
```
how to be sure project.JIBXBinding.unboundJarName will be set (that's why I put it in the afterEvaluate) ?

garc33 replied:
In fact there is no problem. As we have 

```
generateUnboundJar.onlyIf {
    project.JIBXBinding.unboundJarName != null
}
```

if unboundJarName is null, the call to generateUnboundJar.output will be just skipped.

bcorne replied:
Indeed but archiveName won't be correctly set :

```
+        def generateUnboundJar = project.task(type: Jar, dependsOn: 'JIBXResources', 'generateUnboundJar') {
+            classifier = 'nojibxbinding'
+            if (project.JIBXBinding.unboundJarName !=null) {
+                archiveName = project.JIBXBinding.unboundJarName
+            }
+            from project.sourceSets.main.output.files
+        }
```

garc33 replied:
Where is the problem if the archive is not generated ?

Following build lifecycle (http://www.gradle.org/docs/current/userguide/build_lifecycle.html)
- the task is created @ initialization time (with the change)
- During configuration phase, the task is available to configuration other tasks (like artifacts)
- After config the task is disabled by "onlyIf" if unboundJarName is not set
- So the archiveName is never set because closure is not executed @ execution time

bcorne replied:
If the archive is not generated, there's no issue. 

if unboundJarName is set, it will be likely done after applying the plugin in the build.gradle, something like that :

```
apply plugin: 'jibx'

.....


JIBXBinding {
    rootAPIPath = 'my/path/to/bind'
    bindingFiles = ['./path/binding.xml']
    unboundJarName = 'my-unbound-jar.jar'
}
```

The closure configuring the 'generateUnboundJar' task will be executed when the plugin is applied, unboundJarName won't be set yet and so the archiveName won't be too.

The onlyIf part being still in the afterEvaluate block will enable the archive generation since unboundJarName is set at this time but I fear the unbound archive won't have the expected name.

It can be checked easily with this simple gradle build script

```
apply plugin : 'jibx'

task check {
    println "unboundJarName value : $JIBXBinding.unboundJarName"
}

JIBXBinding {
    unboundJarName = 'testvalue.jar'
}
```

calling gradle check won't do anything but during the initialization phase you'll have the following statement :

unboundJarName value : null

garc33 replied:
I got it but it seems (in gradle doc) that plugin Extension are resolves (or mapped) before.
So if you write this (example come from doc):

```
apply plugin: GreetingPlugin

greeting {
    message = 'Hi'
    greeter = 'Gradle'
}

class GreetingPlugin implements Plugin<Project> {
    void apply(Project project) {
        project.extensions.create("greeting", GreetingPluginExtension)
        project.task('hello') << {
            println "${project.greeting.message} from ${project.greeting.greeter}"
        }
    }
}

class GreetingPluginExtension {
    String message
    String greeter
}
```

The result should be:

```
> gradle -q hello
Hi from Gradle
```

http://www.gradle.org/docs/current/userguide/custom_plugins.html (59.3)
About our plugin, I have done some tests and it works fine.

EDIT : 
The task is finalized in this example, I will make the change to ensure the data mapping (even if it seems superfluous).

bcorne replied:
Yes in your case, this is working since you're using the << operator, the closure is executed during task execution.

garc33 committed:
Fix archiveName setting

As generateUnboundJar is now declared during the configuration phase, the unboundJarName parameter should be applied after configuration

garc33 committed:
Override unboundJar name as Jar task

To override unboundJar name, write :
generateUnboundJar{
    archiveName="name.jar"
}
By default [baseName]-[appendix]-[version]-nojibxbinding.jar
